Bug 1853610 - Swallow exceptions from Autofill system service

MatthewTighe opened the pull request:
### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.



### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1853610

jonalmeida commented:
Drive-by 🚗 : Could you please add some details or investigation results into change? It doesn't explain why swallowing the exception is an appropriate solution and would raise some red flags when looking at this in the future.

rahulsainani commented:
Sorry, I missed Jon's comment earlier, please continue the discussion.

MatthewTighe commented:
Ah, my bad. The underlying platform code for `context.getSystemService(AutofillManager::class.java).hasEnabledAutofillServices()` is:

```java
    public boolean hasEnabledAutofillServices() {
        if (mService == null) return false;

        final SyncResultReceiver receiver = new SyncResultReceiver(SYNC_CALLS_TIMEOUT_MS);
        try {
            mService.isServiceEnabled(mContext.getUserId(), mContext.getPackageName(),
                    receiver);
            return receiver.getIntResult() == 1;
        } catch (RemoteException e) {
            throw e.rethrowFromSystemServer();
        } catch (SyncResultReceiver.TimeoutException e) {
            throw new RuntimeException("Fail to get enabled autofill services status.");
        }
    }
```

Hopefully the comment I've added will improve clarity, and I added logging as well. It doesn't seem like there's much we could do here unless we wanted to implement a retry strategy. It would take some investigation to determine why the system service can time out, as the [source](https://android.googlesource.com/platform/prebuilts/fullsdk/sources/android-31/+/refs/heads/androidx-core-core-google-shortcuts-release/com/android/server/autofill/AutofillManagerService.java#1684) isn't immediately helpful. Not sure we'd be able to do anything differently anyway, so this seemed like an appropriate level of fix for a crasher

jonalmeida commented:
Thanks for the explanation! You have an approval already, so feel free to land (once the lint checks pass).

> Hopefully the comment I've added will improve clarity, and I added logging as well.

General advice: a comment should explain the reasoning behind _why_ we wrote that piece of code and not explain the code itself. In this case, your logger message says the same thing as the comment, so it's not needed here (but it's fine to keep as well).

DonalMe commented:
@MatthewTighe there are linter failures in CI https://firefoxci.taskcluster-artifacts.net/RyPZpLpES7-qKYHoH9EKPw/0/public/reports/detekt/detekt.html

MatthewTighe commented:
Whoops, only staged the `fenix` directory with the changes to fix that. Should be good now. Thanks for noticing!

MatthewTighe commented:
@Mergifyio backport releases_v121
